From 1f08e3b3fa2bb25c6ff0a906f49f2a076da4b0e4 Mon Sep 17 00:00:00 2001
From: Barna Zsombor Klara <zsombor.klara@cloudera.com>
Date: Sat, 16 Jul 2016 12:06:37 -0700
Subject: [PATCH 0678/1164] CDH-40038: HIVE-14226 : Invalid check on an
 ASTNode#toStringTree in CalcitePlanner (Barna
 Zsombor Klara via Jesus Camacho Rodriguez)
 Signed-off-by: Ashutosh Chauhan
 <hashutosh@apache.org>

Change-Id: I599329ff3e7d4c1e2524c9d75e586c45ff8ee6b7
---
 .../hadoop/hive/ql/parse/CalcitePlanner.java       |    2 +-
 .../clientnegative/selectDistinctWithoutAggr.q     |    3 +++
 .../clientnegative/selectDistinctWithoutAggr.q.out |    1 +
 3 files changed, 5 insertions(+), 1 deletion(-)
 create mode 100644 ql/src/test/queries/clientnegative/selectDistinctWithoutAggr.q
 create mode 100644 ql/src/test/results/clientnegative/selectDistinctWithoutAggr.q.out

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java b/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java
index f41314c..f2f116f 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java
@@ -2445,7 +2445,7 @@ private RelNode genSelectLogicalPlan(QB qb, RelNode srcRel, RelNode starSrcRel)
               SemanticAnalyzer.unescapeIdentifier(expr.getChild(0).getChild(0).getText()
                   .toLowerCase()), expr, col_list, excludedColumns, inputRR, starRR, pos,
               out_rwsch, tabAliasesForAllProjs, true);
-        } else if (expr.toStringTree().contains("TOK_FUNCTIONDI")
+        } else if (ParseUtils.containsTokenOfType(expr, HiveParser.TOK_FUNCTIONDI)
             && !(srcRel instanceof HiveAggregate)) {
           // Likely a malformed query eg, select hash(distinct c1) from t1;
           throw new CalciteSemanticException("Distinct without an aggreggation.");
diff --git a/ql/src/test/queries/clientnegative/selectDistinctWithoutAggr.q b/ql/src/test/queries/clientnegative/selectDistinctWithoutAggr.q
new file mode 100644
index 0000000..d031513
--- /dev/null
+++ b/ql/src/test/queries/clientnegative/selectDistinctWithoutAggr.q
@@ -0,0 +1,3 @@
+-- Distinct without an aggregation is unsupported
+
+select hash(distinct key) from src;
\ No newline at end of file
diff --git a/ql/src/test/results/clientnegative/selectDistinctWithoutAggr.q.out b/ql/src/test/results/clientnegative/selectDistinctWithoutAggr.q.out
new file mode 100644
index 0000000..2f4bbde
--- /dev/null
+++ b/ql/src/test/results/clientnegative/selectDistinctWithoutAggr.q.out
@@ -0,0 +1 @@
+FAILED: SemanticException 3:7 Distinct keyword is not support in current context. Error encountered near token 'key'
-- 
1.7.9.5

